lib/Dialect/RTL/RTLFolds.cpp

```diff
@@ -706,3 +715,289 @@ void ConcatOp::getCanonicalizationPatterns(OwningRewritePatternList &results,
+}
+
+// Returns the flipped predicate, reversing the LHS and RHS operands.  The
+// lhs and rhs operands should be flipped to match the new predicate.
+static ICmpPredicate flipPredicate(ICmpPredicate predicate) {
+  switch (predicate) {
+  case ICmpPredicate::eq:
+    return ICmpPredicate::eq;
+  case ICmpPredicate::ne:
+    return ICmpPredicate::ne;
+  case ICmpPredicate::slt:
+    return ICmpPredicate::sgt;
+  case ICmpPredicate::sle:
+    return ICmpPredicate::sge;
+  case ICmpPredicate::sgt:
+    return ICmpPredicate::slt;
+  case ICmpPredicate::sge:
+    return ICmpPredicate::sle;
+  case ICmpPredicate::ult:
+    return ICmpPredicate::ugt;
+  case ICmpPredicate::ule:
+    return ICmpPredicate::uge;
+  case ICmpPredicate::ugt:
+    return ICmpPredicate::ult;
+  case ICmpPredicate::uge:
+    return ICmpPredicate::ule;
+  }
+  llvm_unreachable("unknown comparison predicate");
+}
+
+// Try to apply the predicate to an unknown LHS but with a constant RHS.
+// This is accomplished by reasoning about RHS which are on the upper or
+// lower boundary of possible values.
+static IntegerAttr tryApplyCmpPredicateToRHS(ICmpPredicate predicate,
+                                             IntegerAttr &rhs) {
+  MLIRContext *context = rhs.getContext();
+  auto value = rhs.getValue();
+  switch (predicate) {
+  case ICmpPredicate::slt:
+    if (value.isMinSignedValue())
+      return getFalseAttr(context);
+    return {};
+  case ICmpPredicate::sle:
+    if (value.isMaxSignedValue())
+      return getTrueAttr(context);
+    return {};
+  case ICmpPredicate::sgt:
+    if (value.isMaxSignedValue())
+      return getFalseAttr(context);
+    return {};
+  case ICmpPredicate::sge:
+    if (value.isMinSignedValue())
+      return getTrueAttr(context);
+    return {};
+  case ICmpPredicate::ult:
+    if (value.isMinValue())
+      return getFalseAttr(context);
+    return {};
+  case ICmpPredicate::ule:
+    if (value.isMaxValue())
+      return getTrueAttr(context);
+    return {};
+  case ICmpPredicate::ugt:
+    if (value.isMaxValue())
+      return getFalseAttr(context);
+    return {};
+  case ICmpPredicate::uge:
+    if (value.isMinValue())
+      return getTrueAttr(context);
+    return {};
+  default:
+    return {};
+  }
+}
+
+OpFoldResult ICmpOp::fold(ArrayRef<Attribute> constants) {
+
+  // gt a, a -> false
+  // gte a, a -> true
+  if (lhs() == rhs()) {
+    auto val = applyCmpPredicateToEqualOperands(predicate());
+    return IntegerAttr::get(getType(), val);
+  }
+
+  auto lhs = constants.front().dyn_cast_or_null<IntegerAttr>();
+  auto rhs = constants.back().dyn_cast_or_null<IntegerAttr>();
+
+  // gt 1, 2 -> false
+  if (lhs && rhs) {
+    auto val = applyCmpPredicate(predicate(), lhs.getValue(), rhs.getValue());
+    return IntegerAttr::get(getType(), val);
+  }
+
+  // lt x, 0 -> false
+  if (rhs)
+    return tryApplyCmpPredicateToRHS(predicate(), rhs);
+  else if (lhs)
+    return tryApplyCmpPredicateToRHS(flipPredicate(predicate()), lhs);
```

Comment:
Instead of this approach, I'd recommend handling this at canonicalization time: canonicailization would just move any constant to the RHS, and just match it once on that side like you're done.  Then I'd drop it from fold() here.